protected/controllers/PracticeController.php

```diff
@@ -2,62 +2,179 @@
             'accessControl',
-            'ajaxOnly + create',
         );
     }
 
     public function accessRules()
     {
         return array(
             array(
-                'allow',
+                'allow',  // allow users with either the TaskViewPractice or TaskCreatePractice roles to view Practice data
+                'actions' => array('index', 'view'),
+                'roles' => array('TaskViewPractice', 'TaskCreatePractice')
+            ),
+            array(
+                'allow', // allow users with either the TaskCreatePractice or TaskAddPatient roles to perform 'create' actions
                 'actions' => array('create'),
-                'users' => array('@'),
+                'roles' => array('TaskCreatePractice', 'TaskAddPatient'),
+            ),
+            array(
+                'allow', // allow users with the TaskCreatePractice role to perform 'update' actions
+                'actions' => array('update'),
+                'roles' => array('TaskCreatePractice'),
+            ),
+            array(
+                'deny',  // deny all other users
+                'users' => array('*'),
             ),
         );
     }
-    public function actionCreate()
+    /**
+     * Displays a particular model.
+     * @param integer $id the ID of the model to be displayed
+     */
+    public function actionView($id)
+    {
+        $this->render('view', array(
+            'model' => $this->loadModel($id),
+        ));
+    }
+
+    public function actionCreate($context=null)
     {
         $contact = new Contact();
         $address = new Address();
         $practice = new Practice();
 
-        $contact->attributes = $_POST['Contact'];
-        $address->attributes = $_POST['Address'];
-        $practice->attributes = $_POST['Practice'];
-        list($contact,$practice,$address) = $this->performPracticeSave($contact, $practice,$address);
+        $this->performAjaxValidation($contact);
 
-        echo CJSON::encode(array('label'=> $practice->getAddressLines()
-        ,'value'=> $practice->getPrimaryKey()));
+        if (isset($_POST['Contact'])) {
+            $contact->attributes = $_POST['Contact'];
+            $address->attributes = $_POST['Address'];
+            $practice->attributes = $_POST['Practice'];
+            list($contact, $practice, $address) = $this->performPracticeSave($contact, $practice, $address, $context === 'AJAX');
+        }
+        if ($context === 'AJAX') {
+            echo CJSON::encode(array('label'=> $practice->getAddressLines()
+            ,'value'=> $practice->getPrimaryKey()));
+        } else {
+            $this->render('create', array(
+                'model' => $practice,'address' => $address, 'contact' => $contact,
+            ));
+        }
     }
 
-    public function performPracticeSave(Contact $contact, Practice $practice,Address $address)
+    public function performPracticeSave(Contact $contact, Practice $practice,Address $address,$isAjax = false)
```

Comment:
It would be good to have phpdoc for these methods. If you are using phpstorm, right-click anywhere in a file and then click "Generate" then "PHPDoc blocks..." then select the functions you want to add documentation to and click ok, then fill in the blanks